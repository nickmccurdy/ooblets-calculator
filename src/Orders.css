table {
  border: 1px solid;
  border-collapse: collapse;
}

tfoot {
  border-top: 1px solid;
}

td,
th {
  border-left: 1px solid;
  border-right: 1px solid;
}

.fluid {
  width: 100%;
}
