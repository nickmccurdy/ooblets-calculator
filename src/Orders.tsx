import { ArrayHelpers, Field, FormikProps } from "formik"
import { createRef, RefObject, useEffect, useRef } from "react"

import items from "./items"
import "./Orders.css"
import { Order, Values } from "./types"

const defaultOrder: Order = {
  amount: 5,
  item: "caroot",
  payment: 1,
}

function getItem(item: string) {
  return items.flat().find(({ name }) => name === item)!
}

function mapValues<
  K extends string | number | symbol,
  V,
  O extends Record<K, V>,
  VV
>(object: O, callback: (value: V) => VV): Record<K, VV> {
  return Object.fromEntries(
    Object.entries(object).map(([key, value]) => [key, callback(value)])
  )
}

export function calculateOrders(discount: string, orders: Order[]) {
  return orders.map((order) => {
    let orderItems = getItem(order.item)
    if ("from" in orderItems[0]) {
      orderItems = mapValues(orderItems[0].from).map(getItem)
    }

    const cost = orderItems
      .map(
        (item) =>
          ("cost" in item
            ? item.name === discount
              ? Math.floor(item.cost * 0.5)
              : item.cost
            : 0) * order.amount,
      )
      .reduce((x, y) => x + y)
    const profit = order.payment - cost
    // maximum
    const days = orderItems
      .map((item) => ("days" in item ? item.days : 0))
      .reduce((x, y) => Math.max(x, y))
    const profitPerDay = Math.round(profit / (days || 1))

    return { cost, profit, days, profitPerDay }
  })
}

export default function OrdersForm({
  push,
  remove,
  form: { values },
}: ArrayHelpers & { form: FormikProps<Values> }) {
  const inputRefs = useRef<RefObject<HTMLInputElement>[]>([])
  const addingRow = useRef(false)
  const removedRow = useRef<number>()

  useEffect(() => {
    inputRefs.current.push(
      ...values.orders
        .slice(inputRefs.current.length)
        .map(() => createRef<HTMLInputElement>()),
    )

    const last = inputRefs.current[inputRefs.current.length - 1]?.current

    if (last && addingRow.current) {
      last.focus()
      last.select()
      addingRow.current = false
    }

    if (removedRow.current !== undefined) {
      inputRefs.current.splice(removedRow.current, 1)
      removedRow.current = undefined
    }

    if (!values.orders.length) {
      inputRefs.current = []
    }
  })

  const data = calculateOrders(values.discount, values.orders)

  return (
    <table>
      <thead>
        <tr>
          <th>Items</th>
          <th>Cost</th>
          <th>Payment</th>
          <th>Profit</th>
          <th>Days</th>
          <th>Profit/Day</th>
          <th />
        </tr>
      </thead>

      <tbody>
        {data.map((row, index) => {
          function highlightRecommended(
            key: keyof typeof data[0],
            reverse = false,
          ) {
            const results = data
              .map((row, index) => index)
              .sort((a, b) => data[b][key] - data[a][key])

            if (reverse) {
              results.reverse()
            }

            const isRecommended = results.slice(0, values.level).includes(index)
            return isRecommended ? <strong>{row[key]}</strong> : row[key]
          }

          return (
            <tr key={index}>
              <td>
                <Field
                  innerRef={inputRefs.current[index]}
                  name={`orders.${index}.amount`}
                  type="number"
                  step={5}
                  validate={(value: number) => {
                    if (value % 5)
                      return `orders.${index}.amount must by divisible by 5`
                  }}
                  style={{ width: "2.5em" }}
                />

                <Field
                  component="select"
                  name={`orders.${index}.item`}
                  style={{ width: "11em" }}
                >
                  {items.slice(0, values.level).map((items, index) => (
                    <optgroup key={index} label={`Plenny's level ${index + 1}`}>
                      {items.map(({ name }) => (
                        <option key={name}>{name}</option>
                      ))}
                    </optgroup>
                  ))}
                </Field>
              </td>

              <td>{highlightRecommended("cost", true)}</td>

              <td>
                <Field
                  name={`orders.${index}.payment`}
                  type="number"
                  size={5}
                  style={{ width: "4em" }}
                />
              </td>

              <td>{highlightRecommended("profit")}</td>

              <td>{highlightRecommended("days", true)}</td>

              <td>{highlightRecommended("profitPerDay")}</td>

              <td>
                <button
                  type="button"
                  onClick={() => {
                    remove(index)
                    removedRow.current = index
                  }}
                >
                  X
                </button>
              </td>
            </tr>
          )
        })}
      </tbody>

      <tfoot>
        <tr>
          <td colSpan={8}>
            <button
              className="fluid"
              type="button"
              onClick={() => {
                push(defaultOrder)
                addingRow.current = true
              }}
            >
              Add
            </button>
          </td>
        </tr>
      </tfoot>
    </table>
  )
}
