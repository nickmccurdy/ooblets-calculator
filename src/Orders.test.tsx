import { calculateOrders } from "./Orders"

const discount = "lopauliflowers"
const orders = [
  { amount: 20, item: "sweetiebeeties", payment: 286 },
  { amount: 20, item: "planklets", payment: 78 },
  { amount: 15, item: "clothlets", payment: 195 },
  { amount: 10, item: "caroot", payment: 91 },
  { amount: 30, item: "pompadoots", payment: 1248 },
  { amount: 15, item: "springbeans", payment: 144 },
  { amount: 20, item: "muz flour", payment: 192 },
  { amount: 10, item: "lopauliflowers", payment: 384 },
  { amount: 30, item: "hop dobs", payment: 1080 },
  { amount: 20, item: "flooti karioka", payment: 1200 },
  { amount: 15, item: "clambrosia", payment: 600 },
  { amount: 40, item: "oodles", payment: 1920 },
  { amount: 30, item: "pibblepug pies", payment: 5112 },
  { amount: 40, item: "go go cupcakes", payment: 9600 },
  { amount: 40, item: "spressies", payment: 3840 },
  { amount: 40, item: "blue goo pie", payment: 9600 },
]

test("calculates orders", () => {
  expect(calculateOrders(discount, orders)).toMatchSnapshot()
})
