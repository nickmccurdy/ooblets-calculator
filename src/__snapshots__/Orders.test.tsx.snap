// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculates orders 1`] = `
Array [
  Object {
    "cost": 100,
    "days": 3,
    "profit": 186,
    "profitPerDay": 62,
  },
  Object {
    "cost": 0,
    "days": 0,
    "profit": 78,
    "profitPerDay": 78,
  },
  Object {
    "cost": 480,
    "days": 3,
    "profit": -285,
    "profitPerDay": -95,
  },
  Object {
    "cost": 50,
    "days": 1.5,
    "profit": 41,
    "profitPerDay": 27,
  },
  Object {
    "cost": 240,
    "days": 5,
    "profit": 1008,
    "profitPerDay": 202,
  },
  Object {
    "cost": 90,
    "days": 5,
    "profit": 54,
    "profitPerDay": 11,
  },
  Object {
    "cost": 60,
    "days": 2,
    "profit": 132,
    "profitPerDay": 66,
  },
  Object {
    "cost": 50,
    "days": 7,
    "profit": 334,
    "profitPerDay": 48,
  },
  Object {
    "cost": 240,
    "days": 2,
    "profit": 840,
    "profitPerDay": 420,
  },
  Object {
    "cost": 380,
    "days": 4,
    "profit": 820,
    "profitPerDay": 205,
  },
  Object {
    "cost": 75,
    "days": 3,
    "profit": 525,
    "profitPerDay": 175,
  },
  Object {
    "cost": 120,
    "days": 2,
    "profit": 1800,
    "profitPerDay": 900,
  },
  Object {
    "cost": 390,
    "days": 3,
    "profit": 4722,
    "profitPerDay": 1574,
  },
  Object {
    "cost": 1640,
    "days": 5,
    "profit": 7960,
    "profitPerDay": 1592,
  },
  Object {
    "cost": 680,
    "days": 5,
    "profit": 3160,
    "profitPerDay": 632,
  },
  Object {
    "cost": 1520,
    "days": 4,
    "profit": 8080,
    "profitPerDay": 2020,
  },
]
`;
